programs/drift/src/instructions/user.rs

```diff
@@ -2008,3 +2010,467 @@ pub struct DeleteUser<'info> {
+                ctx.accounts.authority.key() == ix.accounts[3].pubkey,
+                ErrorCode::InvalidSwap,
+                "the authority passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.out_spot_market_vault.key() == ix.accounts[4].pubkey,
+                ErrorCode::InvalidSwap,
+                "the out_spot_market_vault passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.in_spot_market_vault.key() == ix.accounts[5].pubkey,
+                ErrorCode::InvalidSwap,
+                "the in_spot_market_vault passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.out_token_account.key() == ix.accounts[6].pubkey,
+                ErrorCode::InvalidSwap,
+                "the out_token_account passed to SwapBegin and End must match"
+            )?;
+
+            validate!(
+                ctx.accounts.in_token_account.key() == ix.accounts[7].pubkey,
+                ErrorCode::InvalidSwap,
+                "the in_token_account passed to SwapBegin and End must match"
+            )?;
+        } else {
+            validate!(
+                ix.program_id == AssociatedToken::id()
+                    || ix.program_id == serum_program::id()
+                    || ix.program_id == jupiter_mainnet_3::ID
+                    || ix.program_id == jupiter_mainnet_4::ID,
+                ErrorCode::InvalidSwap,
+                "only allowed to pass in ixs to ATA or openbook or Jupiter v3 or v4 programs"
+            )?;
+        }
+
+        index += 1;
+    }
+
+    validate!(
+        found_end,
+        ErrorCode::InvalidSwap,
+        "found no SwapEnd instruction in transaction"
+    )?;
+
+    Ok(())
+}
+
+#[access_control(
+    fill_not_paused(&ctx.accounts.state)
+)]
+pub fn handle_end_swap(
+    ctx: Context<Swap>,
+    out_market_index: u16,
+    in_market_index: u16,
+    limit_price: Option<u64>,
+) -> Result<()> {
+    let state = &ctx.accounts.state;
+    let clock = Clock::get()?;
+    let slot = clock.slot;
+    let now = clock.unix_timestamp;
+
+    let AccountMaps {
+        perp_market_map,
+        spot_market_map,
+        mut oracle_map,
+    } = load_maps(
+        &mut ctx.remaining_accounts.iter().peekable(),
+        &MarketSet::new(),
+        &get_writable_spot_market_set_from_many(vec![in_market_index, out_market_index]),
+        clock.slot,
+        Some(state.oracle_guard_rails),
+    )?;
+
+    let user_key = ctx.accounts.user.key();
+    let mut user = load_mut!(&ctx.accounts.user)?;
+
+    let mut user_stats = load_mut!(&ctx.accounts.user_stats)?;
+
+    let mut out_spot_market = spot_market_map.get_ref_mut(&out_market_index)?;
+
+    let out_oracle_data = oracle_map.get_price_data(&out_spot_market.oracle)?;
+    let out_oracle_price = out_oracle_data.price;
+    controller::spot_balance::update_spot_market_cumulative_interest(
```

Comment:
why is this updated after swapping?

Reply:
the deposit/borrow balances aren't updated until below this, so it's actually updated before the swap. could potentially move in to begin_swap if you think that makes more sense